docs/raw/serving-community-history.md

```diff
@@ -0,0 +1,315 @@
+
+Due to the nature of distributed systems, there's no guarantee that a received message is the "last" message. This is especially true when member nodes request historical messages from store nodes. 
+
+Therefore, member nodes MUST wait for 20 seconds after receiving the last `CommunityMessageArchiveIndex` before they start extracting the magnet link to fetch the latest archive index.
+
+### Downloading individual archives
+Once a message archive index is downloaded, community member nodes use a local lookup table to determine which of the listed archives are missing. For this lookup to work, member nodes MUST store the KECCAK-256 hashes of the magnet links for archives they've downloaded.
+
+Given a list of magnet links provided by the index to download individual archives, member nodes can:
+
+1. **Download all archives** - Extract each magnet link in the index and pass them to the underlying BitTorrent client (this is the case for new community member nodes that haven't downloaded any archives yet)
+2. **Download only the latest archive** - Extract only the newest magnet link and pass it to the BitTorrent client (this the case for any member node that already has downloaded all previous history and is now interested in only the latst archive)
+3. **Download specific archives** - Look into `from` and `to` fields of every `WakuMessageArchiveMetadata` and only extract magnet links for archives of a specific time range (can be the case for member nodes that have recently joined the network and are only interested in a subset of the complete history)
+
+
+## Storing historical messages
+
+When message archives are fetched, community member nodes MUST unwrap the resulting `WakuMessage` instances into `ApplicationMetadataMessage` instances and store them in their local database.
+Community member nodes SHOULD NOT store the wrapped `WakuMessage` messages.
+
+Already stored messages with the same `id` or `clock` value MUST be replaced with messages extracted from archives, if both of these values are equal.
+
+Community members nodes MUST ignore the expiration state of each archive message.
+
+## Considerations
+
+The following are things to cosider when implementing this specification.
+
+### Bandwidth consumption
+
+Community member nodes will download the latest archive they've received from the archive index, which includes messages from the last seven days. Assuming that community members nodes were online for that time range, they have already downloaded that message data and will now download an archive that contains the same.
+
+This means there's a possibility member nodes will download the same data at least twice.
+
+### Multiple community owners
+
+It is possible for community owners to export the private key of their owned community and pass it to other users so they become community owners as well. This means, it's possible for multiple owners to exist.
```

Comment:
This is a short term problem that we can probably ignore for now.  The reason we can ignore this is that as soon as possible we will tokenize community ownership with an NFT, and this will ensure that there can only ever be one owner.  Now that one owner *could* run two nodes, but we can detect if this is happening and warn them that they shouldn't do this.  Or if we detected two owner nodes, we could randomly assign only one of the nodes to be able to produce history torrents, and expose a setting for the community owner to select a specific node to hold this responsibility?

Reply:
While we can ignore it, it is still a possible attack vector that at minimum should be mentioned.

It isn't obvious to me that separating the two is going to always be straightforward. Example attack: community owner key gets compromised, and they start posting two different databases that are off-by-one that keeps churning local user db. Not a huge concern and perhaps not very likely, and if community is compromised a user can always leave etc etc.

Reply:
> Or if we detected two owner nodes, we could randomly assign only one of the nodes to be able to produce history torrents, and expose a setting for the community owner to select a specific node to hold this responsibility?

I believe there's no way for us to guarantee/enforce this. Theoretically, community owners could run a version of a node that bypasses all of that and still just publishes magnet links on the special channel.

So I guess the easiest thing we can do to account for that is to set and store a "main owner" that other nodes will then use to verify that the magnet link message was signed by that main owner.

All other magnet link messages will be ignored.